### PR TITLE
Stripe payments

### PR DESCRIPTION
**Pull request by Kovbo**
Do not merge!

**STRIPE SETUP**

Add to .env:
```
STRIPE_PUBLISHABLE_KEY
STRIPE_SECRET_KEY
STRIPE_WEBHOOK_SECRET
```

Setup a new webhook: 
Visit: https://dashboard.stripe.com/webhooks 
Add this API endpoint:  
https://app.openpipe.ai/api/stripe/webhook

Handled events:
```
charge.succeeded
charge.failed
charge.expired
```
payment_intent.succeeded 

We can probably add more events in the stripe dashboard and check for events that we may be interested in. 

We need to create temps of use. Stripe requires:
_Add terms to your website or app that state how you plan to save payment method details and allow customers to opt in. If you want to charge them when they’re offline, make sure your terms include the following:
•	The customer’s agreement to your initiating a payment or a series of payments on their behalf for specified transactions.
•	The anticipated timing and frequency of payments (for example, if the charges are for scheduled installments, subscription payments, or unscheduled top-ups).
•	How you determine the payment amount.
•	Your cancellation policy, if the payment method is for a subscription service.
Make sure you keep a record of your customer’s written agreement to these terms_.

**CRON Jobs**
There are 2 cron jobs:
`generateInvoices.task.ts` - runs at the beginning of each month and creates invoices.
`chargeInvoices.task.ts` - runs once a day. An open question: should we track failed payment attempts and stop trying to charge an account at some point? We could run the job once a month to limit the amount of attempts.

**DB changes**
Added 2 columns:
`project.billable` and `project.stripeCustomerId`

stripeCustomerId connects a stripe user to a project. Billable defines if we issue invoices for a particular project.
We will need to manually set billable=false for some clients.

The current logic for the creation of invoices is:
```
 if (!invoiceAlreadyExists && project.billable && project.id === "89149c5d-aeda-49f7-b668-41104aef8444") {
  await createInvoice(project.id, startOfPreviousMonth, endOfPreviousMonth);
}
```

We can test the creation of invoices and off-site payments on a single project: "SDK test Project".
To backfill invoices we can temporarily change "startOfPreviousMonth" to "startOfPreviousYear".

"Credits backfill"

New projects will automatically receive 100 credits.
To backfill existing projects, use this script: `backfillCredits.ts`
It adds 100 credits to all the existing projects.

### CODE REVIEW

File: app/src/components/billing/BillingContentTabs/PaymentMethods/AddPaymentMethodButton.tsx
Comment (arcticfly) [inline app/src/components/billing/BillingContentTabs/PaymentMethods/AddPaymentMethodButton.tsx]: Should this go in our environment variables?

File: app/src/components/billing/BillingContentTabs/PaymentMethods/AddPaymentMethodButton.tsx
Comment (arcticfly) [inline app/src/components/billing/BillingContentTabs/PaymentMethods/AddPaymentMethodButton.tsx]: To stay consistent with Chakra styling, use `Box` here instead.

File: app/src/components/billing/BillingContentTabs/PaymentMethods/PaymentMethods.tsx
Comment (arcticfly) [inline app/src/components/billing/BillingContentTabs/PaymentMethods/PaymentMethods.tsx]: Nit: rename `isDefaultChangeLoading` to something like `setDefaultInProgress`

File: app/src/components/billing/BillingContentTabs/PaymentMethods/PaymentMethods.tsx
Comment (arcticfly) [inline app/src/components/billing/BillingContentTabs/PaymentMethods/PaymentMethods.tsx]: Nit: Change `isDeleteLoading` to something like `deletionInProgress`

File: app/src/components/billing/BillingContentTabs/PaymentMethods/PaymentMethods.tsx
Comment (arcticfly) [inline app/src/components/billing/BillingContentTabs/PaymentMethods/PaymentMethods.tsx:87]: Instead of adding custom <Spinner /> elements in loading buttons, you can use Chakra's built-in `isLoading` prop. That should also disable the button while it's loading.

File: app/src/pages/api/stripe/webhook.ts
Comment (arcticfly) [inline app/src/pages/api/stripe/webhook.ts:48]: This seems like a good place to check if the invoiceId exists. That way we can change the function signature of all the functions you call later to expect that invoiceId is a string, not string or undefined.

File: app/src/pages/api/stripe/webhook.ts
Comment (arcticfly) [inline app/src/pages/api/stripe/webhook.ts]: Since we only use findProject in one place, I don't think we need to define it as a separate function (unless you plan to use it in more places later?)

Also, I think we should capture an exception in Sentry if we aren't able to find an invoice or a project. Search the codebase for `captureException` to see how we do that.

**Review by arcticfly — COMMENTED** [review]
I haven't fully reviewed, but I want to avoid conflicts with @corbt 😆

File: app/src/components/billing/BillingContentTabs/Invoices/InvoicesTable.tsx
Comment (corbt) [inline app/src/components/billing/BillingContentTabs/Invoices/InvoicesTable.tsx]: There's actually an `isLoading` prop on buttons that handles this. I usually also set `isDisabled` to true when it's processing to prevent someone from accidentally double-clicking the button.

File: app/src/components/billing/BillingContentTabs/PaymentMethods/AddPaymentMethodButton.tsx
Comment (Kovbo) [reply app/src/components/billing/BillingContentTabs/PaymentMethods/AddPaymentMethodButton.tsx]: Oh, I was testing and forgot to replace it!

File: app/src/server/utils/stripe.ts
Comment (arcticfly) [inline app/src/server/utils/stripe.ts:126]: Good use of overloaded function signatures.

File: app/src/server/utils/stripe.ts
Comment (arcticfly) [inline app/src/server/utils/stripe.ts:98]: I would check to ensure that project exists here and throw a 404 if it doesn't. That way you don't need all the `project?.` later in the function, and we don't create customers for nonexistent projects.

File: app/src/server/utils/stripe.ts
Comment (arcticfly) [inline app/src/server/utils/stripe.ts]: Nit: add space between // and Associate

File: app/src/server/emails/sendPaymentFailed.ts
Comment (corbt) [inline app/src/server/emails/sendPaymentFailed.ts]: Let's replace the first sentence with the following: "We attempted to charge your default payment method for project ${projectName} but the payment failed."

File: app/src/server/emails/sendPaymentFailed.ts
Comment (corbt) [inline app/src/server/emails/sendPaymentFailed.ts:1]: We should also have an email draft for when the payment went through successfully.

File: app/src/server/api/routers/payments.router.ts
Comment (arcticfly) [inline app/src/server/api/routers/payments.router.ts:49]: Under what circumstances will the stripe user be deleted?

File: app/src/server/api/routers/payments.router.ts
Comment (corbt) [inline app/src/server/api/routers/payments.router.ts]: I don't fully grok the flow yet, but will `createStripeCustomer` and `createStripeIntent` always be called at the same time? If so, should they be combined into a single mutation?

File: app/src/server/utils/stripe.ts
Comment (arcticfly) [inline app/src/server/utils/stripe.ts]: You could make these functions even more streamlined with arrow function syntax:

```
export const createSeupIntent = async (stripeCustomerId: string) =>
stripe.setupIntents.create({
...
```

File: app/src/components/billing/BillingContentTabs/PaymentMethods/AddPaymentMethodButton.tsx
Comment (Kovbo) [reply app/src/components/billing/BillingContentTabs/PaymentMethods/AddPaymentMethodButton.tsx]: Yes, fixed!

File: app/src/server/tasks/chargeInvoices.task.ts
Comment (corbt) [inline app/src/server/tasks/chargeInvoices.task.ts]: Does this link work? It looks like "invoices" has a typo at a minimum.

File: app/src/components/billing/BillingContentTabs/PaymentMethods/PaymentMethods.tsx
Comment (Kovbo) [reply app/src/components/billing/BillingContentTabs/PaymentMethods/PaymentMethods.tsx:87]: So convenient, I did not know it! Changed in all the places.

File: app/src/server/tasks/chargeInvoices.task.ts
Comment (corbt) [inline app/src/server/tasks/chargeInvoices.task.ts]: "Payment" spelling

File: app/src/server/tasks/chargeInvoices.task.ts
Comment (corbt) [inline app/src/server/tasks/chargeInvoices.task.ts]: "Payment successfully applied"

File: app/src/server/api/routers/payments.router.ts
Comment (arcticfly) [inline app/src/server/api/routers/payments.router.ts]: Nit: `defaultPaymentMethodId === undefined` -> `!defaultPaymentMethodId`

File: app/src/server/api/routers/payments.router.ts
Comment (arcticfly) [inline app/src/server/api/routers/payments.router.ts]: Remove

File: app/src/server/tasks/chargeInvoices.task.ts
Comment (arcticfly) [inline app/src/server/tasks/chargeInvoices.task.ts:37]: If the invoice status is ERROR (due to an earlier payment method being declined) shouldn't we let the user pay in this function?

Comment (arcticfly) [reply app/src/server/tasks/chargeInvoices.task.ts:37]: I'd also prefer not to check the amount of the invoice here. If the invoice is below a certain threshold, we should just mark it as cancelled when we create it.

File: app/src/pages/api/stripe/webhook.ts
Comment (Kovbo) [reply app/src/pages/api/stripe/webhook.ts:48]: For some reason, I could not do it. returning and breaking the switch/case did not help with typings. Let me take a look again.

File: app/src/server/tasks/chargeInvoices.task.ts
Comment (arcticfly) [inline app/src/server/tasks/chargeInvoices.task.ts]: Since we may actually set this variable to a non-default payment method, let's rename it to something like `paymentMethodToUse`

File: app/src/server/tasks/chargeInvoices.task.ts
Comment (arcticfly) [inline app/src/server/tasks/chargeInvoices.task.ts]: Shouldn't this error be "Add a payment method."?

File: app/src/pages/api/stripe/webhook.ts
Comment (Kovbo) [reply app/src/pages/api/stripe/webhook.ts]: We might need to add other event handlers and send notifications about successful payments. We would need a separate function in this case, but I will remove it for now.

Added `captureException,` but I am unsure if I did it correctly, since the codebase has different implementations.

File: app/src/server/tasks/chargeInvoices.task.ts
Comment (arcticfly) [inline app/src/server/tasks/chargeInvoices.task.ts]: Do you have to include this `?? "app.openpipe.com"` logic? I think we require `NEXT_PUBLIC_HOST` to be set.

File: app/src/server/tasks/generateInvoices.task.ts
Comment (arcticfly) [inline app/src/server/tasks/generateInvoices.task.ts]: Nit: add space between // and the rest of comments to improve readability.

File: app/src/server/api/routers/projects.router.ts
Comment (corbt) [inline app/src/server/api/routers/projects.router.ts]: Actually let's only do this for the first default project we create when a user signs up, instead of for every new project an existing user creates.

File: app/src/server/tasks/generateInvoices.task.ts
Comment (arcticfly) [inline app/src/server/tasks/generateInvoices.task.ts]: Nit: I would use "generate" instead of "get" in the names of these functions. "get" implies that we're retrieving information that already exists, whereas "generate" implies that we're creating a new value.

File: app/src/server/tasks/worker.ts
Comment (arcticfly) [inline app/src/server/tasks/worker.ts]: I think we should only run this once a month to avoid overdraft fees on our users' payment methods if they don't have sufficient funds.

File: app/src/server/utils/emails.ts
Comment (arcticfly) [inline app/src/server/utils/emails.ts:1]: I would move this file into src/server/emails and rename if something specific to its task like `emailAdminsAboutPaymentFailure.ts`.

**Review by arcticfly — COMMENTED** [review]
Nice. You've already thought through a variety of tricky edge cases.

**Review by corbt — COMMENTED** [review]
Looks pretty good overall!

File: app/src/server/utils/stripe.ts
Comment (Kovbo) [reply app/src/server/utils/stripe.ts:98]: Completely agree; changed it.

File: app/src/server/api/routers/payments.router.ts
Comment (Kovbo) [reply app/src/server/api/routers/payments.router.ts:49]: A user can be deleted through the stripe dashboard. In this case, it would throw an error.
A stripe user has a type of `Stripe.Response<Stripe.Customer | Stripe.DeletedCustomer>`.

File: app/src/server/utils/stripe.ts
Comment (Kovbo) [reply app/src/server/utils/stripe.ts]: Yes, it is a good place for an arrow function.

File: app/src/server/tasks/chargeInvoices.task.ts
Comment (Kovbo) [reply app/src/server/tasks/chargeInvoices.task.ts:37]: Currently, an invoice would have a "PENDING" status if payment was declined. This way, the user can update payment details and pay it manually, or the system will charge automatically once the corn job executes.

When creating an invoice, I cancel it by default if the amount is <= 1. We can probably remove this check, but since the payment logic is different from the invoice creation logic, I explicitly added it.

File: app/src/server/tasks/chargeInvoices.task.ts
Comment (Kovbo) [reply app/src/server/tasks/chargeInvoices.task.ts]: Nice idea, changed.

File: app/src/server/tasks/chargeInvoices.task.ts
Comment (Kovbo) [reply app/src/server/tasks/chargeInvoices.task.ts]: Yes, good idea.

File: app/src/server/tasks/chargeInvoices.task.ts
Comment (Kovbo) [reply app/src/server/tasks/chargeInvoices.task.ts]: Oh yes, I wrote it before I discovered the `env.mjs` helper :)

File: app/src/server/tasks/generateInvoices.task.ts
Comment (Kovbo) [reply app/src/server/tasks/generateInvoices.task.ts]: I see, nice idea!

File: app/src/server/tasks/worker.ts
Comment (Kovbo) [reply app/src/server/tasks/worker.ts]: Yes, it may be a solution since we have a manual "pay" button.
I guess we can create invoices on the 1st day of the month and charge clients on the second?

File: app/src/server/utils/emails.ts
Comment (Kovbo) [reply app/src/server/utils/emails.ts:1]: Ok, got it! I considered the emails folder as a collection of email templates. I changed the approach a bit. I created a function `sendToAdmins` that accepts a specific email template as a callback.

File: app/src/components/billing/BillingContentTabs/Invoices/InvoicesTable.tsx
Comment (Kovbo) [reply app/src/components/billing/BillingContentTabs/Invoices/InvoicesTable.tsx]: Oh yeah, I just discovered it! This way, the code looks much better.

File: app/src/server/emails/sendPaymentFailed.ts
Comment (Kovbo) [reply app/src/server/emails/sendPaymentFailed.ts]: Okay, will surely do it!

File: app/src/server/emails/sendPaymentFailed.ts
Comment (Kovbo) [reply app/src/server/emails/sendPaymentFailed.ts:1]: Will take a look at it.

File: app/src/server/tasks/chargeInvoices.task.ts
Comment (Kovbo) [reply app/src/server/tasks/chargeInvoices.task.ts]: Oh, yes, sorry. Actually the `returnUrl` is redundant here as we use off-site payments without redirections, but for some reason stripe includes it in docs.

File: app/src/server/emails/sendPaymentFailed.ts
Comment (Kovbo) [reply app/src/server/emails/sendPaymentFailed.ts:1]: Added payment successful emails.

File: app/src/server/api/routers/payments.router.ts
Comment (Kovbo) [reply app/src/server/api/routers/payments.router.ts]: Yes, it should work fine!

File: app/src/server/api/routers/projects.router.ts
Comment (Kovbo) [reply app/src/server/api/routers/projects.router.ts]: Sure, changed it!

File: app/src/server/tasks/worker.ts
Comment (arcticfly) [reply app/src/server/tasks/worker.ts]: Ideally not too much time would pass between the creation of invoices and the time we charge them. Could we create the invoices at midnight, then charge them a few hours later, perhaps around 8am UTC?

File: app/src/server/utils/emails.ts
Comment (arcticfly) [reply app/src/server/utils/emails.ts:1]: I don't currently see that function in this PR. Have you just not pushed it yet?

**Review by corbt — APPROVED** [review]
Ok, I think this is good to merge once the merge conflicts are fixed. Nice work!

File: app/src/pages/api/stripe/webhook.ts
Comment (arcticfly) [reply app/src/pages/api/stripe/webhook.ts:48]: Ah, I think I see the problem. You need to define invoiceId before the switch statement instead of in one of its conditional blocks.

File: app/src/server/utils/emails.ts
Comment (arcticfly) [reply app/src/server/utils/emails.ts:1]: My mistake, I was looking at an old version of the code.

File: app/src/server/tasks/worker.ts
Comment (Kovbo) [reply app/src/server/tasks/worker.ts]: Good idea, I will change it once we test the production payments!

**Pull request merged**
